File: ios/Classes/BackgroundTaskPlugin.swift
/**
Copyright [2024] [Never Inc.]
Copyright [2019] [Ali Almoullim]

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import Flutter
import UIKit
import CoreLocation

public class BackgroundTaskPlugin: NSObject, FlutterPlugin, CLLocationManagerDelegate {

    public static let dispatchEngine: FlutterEngine = FlutterEngine(
        name: Bundle.main.bundleIdentifier ?? "background_task",
        project: nil,
        allowHeadlessExecution: true
    )
    public static var onRegisterDispatchEngine: (() -> Void)?
    static var isRegisteredDispatchEngine = false
    
    static var locationManager: CLLocationManager?
    static var channel: FlutterMethodChannel?
    static var statusEventChannel: FlutterEventChannel?
    static var beaconEventChannel: FlutterEventChannel?
    static var isRunning = false
    
    static var dispatchChannel: FlutterMethodChannel?
    static var dispatcherRawHandle: Int?
    static var handlerRawHandle: Int?
    
    static var beaconRegion:CLBeaconRegion?
    static var UUIDList = [
        "D30A3941-35F9-D31A-215B-1EACF2DADB8B"//TODO: デフォルトUUID
    ]
    
    var beacons: [Any] = []
    
    private var isEnabledEvenIfKilled: Bool {
        return UserDefaultsRepository.instance.fetchIsEnabledEvenIfKilled()
    }
    
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let instance = BackgroundTaskPlugin()
        registrar.addApplicationDelegate(instance)
        
        let channel = FlutterMethodChannel(name: ChannelName.methods.value, binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(instance, channel: channel)
        channel.setMethodCallHandler(instance.handle)
        BackgroundTaskPlugin.channel = channel
        
        let statusEventChannel = FlutterEventChannel(name: ChannelName.statusEvent.value, binaryMessenger: registrar.messenger())
        statusEventChannel.setStreamHandler(StatusEventStreamHandler())
        BackgroundTaskPlugin.statusEventChannel = statusEventChannel
        
        let beaconEventChannel = FlutterEventChannel(name: ChannelName.beaconEvent.value, binaryMessenger: registrar.messenger())
        beaconEventChannel.setStreamHandler(BeaconEventStreamHandler())
        BackgroundTaskPlugin.beaconEventChannel = beaconEventChannel
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        if (call.method == "is_running_background_task") {
            result(Self.isRunning)
        } else if (call.method == "set_background_handler") {
            let args = call.arguments as? Dictionary<String, Any>
            Self.dispatcherRawHandle = args?["callbackDispatcherRawHandle"] as? Int
            Self.handlerRawHandle = args?["callbackHandlerRawHandle"] as? Int
            debugPrint("registered \(String(describing: args))")
            result(true)
        } else if (call.method == "start_beacon_task"){
            let args = call.arguments as? Dictionary<String, Any>
            let setUuid = (args?["uuid"] as? String) ?? ""
            let isEnabledEvenIfKilled = (args?["isEnabledEvenIfKilled"] as? Bool) ?? false
            Self.UUIDList[0] = setUuid
            
            let userDefaultsRepository = UserDefaultsRepository.instance
            userDefaultsRepository.removeRawHandle()
            if let dispatcherRawHandle = Self.dispatcherRawHandle, let handlerRawHandle = Self.handlerRawHandle {
                userDefaultsRepository.save(
                    callbackDispatcherRawHandle: dispatcherRawHandle,
                    callbackHandlerRawHandle: handlerRawHandle
                )
            }
            userDefaultsRepository.saveIsEnabledEvenIfKilled(isEnabledEvenIfKilled)
            
            registerDispatchEngine()
            
            let locationManager = CLLocationManager()
            locationManager.delegate = self
            locationManager.requestAlwaysAuthorization()
            
            for val in Self.UUIDList{
                let uuid: NSUUID! = NSUUID(uuidString: val.lowercased())
                let identifierStr = "No1"
                Self.beaconRegion = CLBeaconRegion(uuid: uuid as UUID, identifier: identifierStr)
                Self.beaconRegion!.notifyEntryStateOnDisplay = true;
                Self.beaconRegion!.notifyOnEntry = true;
                Self.beaconRegion!.notifyOnExit = true;
                locationManager.startMonitoring(for: Self.beaconRegion!)
                print("見つけます:" + val)
            }
            Self.locationManager = locationManager
            Self.isRunning = true
            StatusEventStreamHandler.eventSink?(
                StatusEventStreamHandler.StatusType.start(message: "beacon").value
            )
            result(true)
        } else if (call.method == "stop_beacon_task"){
            for val in Self.UUIDList{
                Self.locationManager!.stopMonitoring(for: Self.beaconRegion!)
            }
            Self.isRunning = false
            StatusEventStreamHandler.eventSink?(
                StatusEventStreamHandler.StatusType.stop.value
            )
            result(true)
        }
    }
    
    public func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [AnyHashable : Any] = [:]) -> Bool {
        return true
    }

    public func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
        let isEnabled = status == .authorizedAlways || status == .authorizedWhenInUse
        StatusEventStreamHandler.eventSink?(
            StatusEventStreamHandler.StatusType.permission(message: "\(isEnabled ? "enabled" : "disabled")").value
        )
    }

    public func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        debugPrint("didFailWithError: \(error)")
        StatusEventStreamHandler.eventSink?(
            StatusEventStreamHandler.StatusType.error(message: error.localizedDescription).value
        )
    }
    
    private func registerDispatchEngine() {
        if (Self.isRegisteredDispatchEngine) {
            return
        }
        let handle = UserDefaultsRepository.instance.fetchCallbackDispatcherRawHandle()
        if let info = FlutterCallbackCache.lookupCallbackInformation(Int64(handle)) {
            Self.dispatchEngine.run(withEntrypoint: info.callbackName, libraryURI: info.callbackLibraryPath)
            Self.onRegisterDispatchEngine?()
            let dispatchChannel = FlutterMethodChannel(
                name: ChannelName.methods.value,
                binaryMessenger: Self.dispatchEngine.binaryMessenger
            )
            dispatchChannel.setMethodCallHandler { call, result in
                if (call.method == "callback_channel_initialized") {
                    dispatchChannel.invokeMethod("notify_callback_dispatcher", arguments: nil)
                    result(true)
                }
            }
            Self.dispatchChannel = dispatchChannel
            Self.isRegisteredDispatchEngine = true
        }
    }
    
    private func sendData(event: ServiceEvents, value: [String : Any?]){
        let callbackHandlerRawHandle = UserDefaultsRepository.instance.fetchCallbackHandlerRawHandle()
        let data = [
            "callbackHandlerRawHandle": callbackHandlerRawHandle,
            "data": value
        ] as [String : Any?]
        Self.dispatchChannel?.invokeMethod(event.getName(), arguments: data)
        
        
        if(event == ServiceEvents.Monitor){
            BeaconEventStreamHandler.eventSink?(
                value
            )
        }
    }
    
    public func locationManager(_ manager: CLLocationManager, didStartMonitoringFor region: CLRegion) {
        print("見つけ:didStartMonitoringFor")
        manager.requestState(for: region);
    }
    
    public func locationManager(_ manager: CLLocationManager, didDetermineState state: CLRegionState, for region: CLRegion) {
        switch (state) {
        case .inside:
            print("見つかり:inside")
            self.sendData(event: ServiceEvents.Monitor, value:                 [
                "region" : (region as! CLBeaconRegion).uuid.uuidString,
                "state"  : 1
            ])
            break;
        case .outside:
            print("見つから:outside")
            self.sendData(event: ServiceEvents.Monitor, value:                 [
                "region" : (region as! CLBeaconRegion).uuid.uuidString,
                "state"  : 0
            ])
            break;
        case .unknown:
            print("iBeacon unknown")
            break;
        }
    }
    
    public func locationManager(_ manager: CLLocationManager, didEnterRegion region: CLRegion) {
        print("見つかり:didEnterRegion")
        self.sendData(event: ServiceEvents.Monitor, value:                 [
            "region" : (region as! CLBeaconRegion).uuid.uuidString,
            "state"  : 1
        ])
    }
    
    public func locationManager(_ manager: CLLocationManager, didExitRegion region: CLRegion) {
        print("見つから:didExitRegion")
        self.sendData(event: ServiceEvents.Monitor, value:                 [
            "region" : (region as! CLBeaconRegion).uuid.uuidString,
            "state"  : 0
        ])
    }
}

File: android/src/main/kotlin/com/neverjp/background_task/BackgroundTaskPlugin.kt
/**
Copyright [2024] [Never Inc.]
Copyright [2019] [Ali Almoullim]

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package com.neverjp.background_task

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.os.Build
import android.util.Log
import androidx.core.app.ActivityCompat
import androidx.annotation.RequiresApi
import androidx.lifecycle.Observer
import com.neverjp.background_task.lib.BeaconEventStreamHandler
import com.neverjp.background_task.lib.ChannelName
import com.neverjp.background_task.lib.StatusEventStreamHandler
import io.flutter.plugin.common.PluginRegistry
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import org.altbeacon.beacon.Beacon

/** BackgroundTaskPlugin */
class BackgroundTaskPlugin: FlutterPlugin, MethodCallHandler, ActivityAware, PluginRegistry.RequestPermissionsResultListener {

  private var context: Context? = null
  private lateinit var channel : MethodChannel
  private var activity: Activity? = null
  private var statusEventChannel: EventChannel? = null
  private var beaconEventChannel: EventChannel? = null
  private var dispatcherRawHandle: Long? = null
  private var handlerRawHandle: Long? = null
  private val isEnabledEvenIfKilled: Boolean
    get() = pref.getBoolean(BeaconService.isEnabledEvenIfKilledKey, false)
  private val pref: SharedPreferences
    get() =  context!!.getSharedPreferences(BeaconService.PREF_FILE_NAME, Context.MODE_PRIVATE)

  companion object {
    private val TAG = BackgroundTaskPlugin::class.java.simpleName
    private const val REQUEST_PERMISSIONS_REQUEST_CODE = 34
  }

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    val messenger = flutterPluginBinding.binaryMessenger
    context = flutterPluginBinding.applicationContext
    channel = MethodChannel(messenger, ChannelName.METHODS.value)
    channel.setMethodCallHandler(this)

    statusEventChannel = EventChannel(messenger, ChannelName.STATUS_EVENT.value)
    statusEventChannel?.setStreamHandler(StatusEventStreamHandler())

    beaconEventChannel = EventChannel(messenger, ChannelName.BEACON_EVENT.value)
    beaconEventChannel?.setStreamHandler(BeaconEventStreamHandler())
  }

  @RequiresApi(Build.VERSION_CODES.O)
  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
        "set_android_notification" -> {
          setAndroidNotification(call.argument("title"),call.argument("message"),call.argument("icon"))
          result.success(true)
        }
        "is_running_background_task" -> {
          result.success(BeaconService.isRunning)
        }
        "callback_channel_initialized" -> {
          channel.invokeMethod("notify_callback_dispatcher", null)
        }
        "set_background_handler" -> {
          dispatcherRawHandle = call.argument<Long>(BeaconService.callbackDispatcherRawHandleKey)
          handlerRawHandle = call.argument<Long>(BeaconService.callbackHandlerRawHandleKey)
          Log.d(TAG, "registered ${call.arguments}")
          result.success(true)
        }
        "start_beacon_task" -> {
          val distanceFilter = call.argument<Double>(BeaconService.distanceFilterKey)
          val isEnabledEvenIfKilled = call.argument<Boolean>("isEnabledEvenIfKilled") ?: false
          val uuid = call.argument<String>("uuid") ?: ""

          pref.edit().apply {
            remove(BeaconService.callbackDispatcherRawHandleKey)
            remove(BeaconService.callbackHandlerRawHandleKey)
            if (dispatcherRawHandle != null && handlerRawHandle != null) {
              putLong(BeaconService.callbackDispatcherRawHandleKey, dispatcherRawHandle ?: 0)
              putLong(BeaconService.callbackHandlerRawHandleKey, handlerRawHandle ?: 0)
            }
            putFloat(BeaconService.distanceFilterKey, distanceFilter?.toFloat() ?: 0.0.toFloat())
            putBoolean(BeaconService.isEnabledEvenIfKilledKey, isEnabledEvenIfKilled)
          }.apply()
          startBeaconService(uuid)
          result.success(true)
        }
        "stop_beacon_task" ->{
          stopBeaconService()
          result.success(true)
        }
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
    activity = binding.activity
    binding.addRequestPermissionsResultListener(this)
  }

  override fun onDetachedFromActivityForConfigChanges() {
    this.onDetachedFromActivity()
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
    this.onAttachedToActivity(binding)
  }

  override fun onDetachedFromActivity() {
    if (isEnabledEvenIfKilled) {
      BeaconService.statusLiveData.removeObserver(statusObserver)
      BeaconService.beaconLiveData.removeObserver(beaconObserver)
    } else {
      stopBeaconService()
    }
  }

  override fun onRequestPermissionsResult(
    requestCode: Int,
    permissions: Array<out String>,
    grantResults: IntArray
  ): Boolean {
    if (requestCode == REQUEST_PERMISSIONS_REQUEST_CODE) {
      when (PackageManager.PERMISSION_GRANTED) {
          grantResults[0] -> {
            StatusEventStreamHandler.eventSink?.success(
              StatusEventStreamHandler.StatusType.Permission("enabled").value
            )
          }
          else ->  {
            StatusEventStreamHandler.eventSink?.success(
              StatusEventStreamHandler.StatusType.Permission("disabled").value
            )
            Log.d(TAG, "permission is denied")
          }
      }
    }
    return true
  }

  private val statusObserver = Observer<String> {
    StatusEventStreamHandler.eventSink?.success(it)
  }

  private val beaconObserver = Observer<HashMap<String, Any?>> {
    BeaconEventStreamHandler.eventSink?.success(it)
  }

  private fun startBeaconService(uuid: String) {
    if (!checkPermissions()) {
      requestPermissions()
    }

    val intent = Intent(context, BeaconService::class.java)
    context!!.stopService(intent)

    BeaconService.beaconLiveData.observeForever(beaconObserver)
    BeaconService.statusLiveData.observeForever(statusObserver)

    intent.putExtra("uuid", uuid)

    context!!.startService(intent)
  }

  private fun stopBeaconService() {
    val intent = Intent(context, BeaconService::class.java)
    context!!.stopService(intent)
    BeaconService.statusLiveData.value = StatusEventStreamHandler.StatusType.Stop.value
    BeaconService.statusLiveData.removeObserver(statusObserver)
    BeaconService.beaconLiveData.removeObserver(beaconObserver)
  }

  private fun checkPermissions(): Boolean {
    return PackageManager.PERMISSION_GRANTED == ActivityCompat.checkSelfPermission(context!!, Manifest.permission.ACCESS_FINE_LOCATION)
  }

  private fun requestPermissions() {
    activity?.also {
      val shouldProvideRationale = ActivityCompat.shouldShowRequestPermissionRationale(it, Manifest.permission.ACCESS_FINE_LOCATION)
      if (!shouldProvideRationale) {
        ActivityCompat.requestPermissions(it,
          arrayOf(Manifest.permission.ACCESS_FINE_LOCATION),
          REQUEST_PERMISSIONS_REQUEST_CODE)
      }
    }
  }

  private fun setAndroidNotification(title: String?, message: String?, icon: String?) {
    if (title != null) BeaconService.NOTIFICATION_TITLE = title
    if (message != null) BeaconService.NOTIFICATION_MESSAGE = message
    if (icon != null) BeaconService.NOTIFICATION_ICON = icon
  }
}


